src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs

```diff
@@ -1,28 +1,125 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Buffers.Binary;
 using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
 using System.Reflection;
-using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 using System.Runtime.Serialization;
 using System.Runtime.Versioning;
-using System.Threading;
 
 namespace System.Runtime.CompilerServices
 {
     public static partial class RuntimeHelpers
     {
         [Intrinsic]
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        public static extern void InitializeArray(Array array, RuntimeFieldHandle fldHandle);
+        public static unsafe void InitializeArray(Array array, RuntimeFieldHandle fldHandle)
+        {
+            if (array is null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern unsafe void* GetSpanDataFrom(
+            if (fldHandle.IsNullHandle())
+                throw new ArgumentException(SR.Argument_InvalidHandle);
+
+            IRuntimeFieldInfo fldInfo = fldHandle.GetRuntimeFieldInfo();
+
+            if (!RuntimeFieldHandle.GetRVAFieldInfo(fldInfo.Value, out IntPtr address, out uint size))
+                throw new ArgumentException(SR.Argument_BadFieldForInitializeArray);
+
+            // Note that we do not check that the field is actually in the PE file that is initializing
+            // the array. Basically, the data being published can be accessed by anyone with the proper
+            // permissions (C# marks these as assembly visibility, and thus are protected from outside
+            // snooping)
+
+            MethodTable* pMT = GetMethodTable(array);
+            TypeHandle elementTH = pMT->GetArrayElementTypeHandle();
+
+            if (elementTH.IsTypeDesc || !elementTH.AsMethodTable()->IsPrimitive) // Enum is included
+                throw new ArgumentException(SR.Argument_BadArrayForInitializeArray);
+
+            nuint totalSize = pMT->ComponentSize * array.NativeLength;
+
+            // make certain you don't go off the end of the rva static
+            if (totalSize > size)
+                throw new ArgumentException(SR.Argument_BadFieldForInitializeArray);
+
+            ref byte src = ref *(byte*)address;
+            GC.KeepAlive(fldInfo);
```

Comment:
Why is this sufficient here and not later on? I don't think the ref above is going to keep the `fldInfo` alive.

Reply:
It just needs to keep the containing assembly alive. `fldInfo` is not used later.

Reply:
What is keeping the assembly alive from here on? We should be focusing on type lifetimes, not assemblies. I say that because I don't see an occurence of the assembly here, but rather just the type. The above line isn't going to extend the lifetime of the type, so I'm not sure what this is accomplishing. Perhaps we need the lifetime extended in the caller?

Reply:
> The above line isn't going to extend the lifetime of the type

The above line is going to extend the lifetime of the assembly (and type) that contains the backing static field. (Look for https://github.com/search?q=repo%3Adotnet%2Fruntime%20AssociateMemoryWithLoaderAllocator&type=code if you are interested in how this works.)

Reply:
So it looks like `GetAssociatedLoaderAllocator_Unsafe` is the check used by the GC to extend RVA static lifetimes. Can we get a comment to the effect of:

`// RVA static refs extend the lifetime of the static`

Refs, to my understanding, don't generally do this so it is a bit surprising this handles that case and so I think it is appropriate. If we've baked this assumption in other places, then so be it. I would argue in most cases all new `GC.KeepAlive()` calls should be documented.